# 기술 통계 ~

In [28]:
import pandas as pd
new_df = pd.read_csv('K-League-data2.csv')
new_df.info()

# 각 컬렴 별 타입과 개수, null 값이 있는 지 알 수 있음
# ex) Rnd. / 666개 / null값 없음 / Data type : int64
# ex) dtype : 소수 38개, 정수 47개, 오브젝트 5개
# ( 666, 100 )

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 666 entries, 0 to 665
Data columns (total 100 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Unnamed: 0   666 non-null    int64  
 1   Rnd.         666 non-null    int64  
 2   대회           666 non-null    object 
 3   홈 득점         666 non-null    int64  
 4   홈 도움         666 non-null    int64  
 5   홈 슈팅         666 non-null    int64  
 6   홈 유효 슈팅      666 non-null    int64  
 7   홈 블락된슈팅      666 non-null    int64  
 8   홈 벗어난슈팅      666 non-null    int64  
 9   홈 PA내 슈팅     666 non-null    int64  
 10  홈 PA외 슈팅     666 non-null    int64  
 11  홈 프리킥 슈팅     666 non-null    int64  
 12  홈 프리킥 유효슈팅   666 non-null    int64  
 13  홈 프리킥 크로스    666 non-null    float64
 14  홈 오프사이드      666 non-null    int64  
 15  홈 코너킥        666 non-null    int64  
 16  홈 스로인        666 non-null    int64  
 17  홈 드리블        666 non-null    float64
 18  홈 패스         666 non-null    float64
 19  홈 키패스  

K-League-data2.csv 데이터는 가공을 한번 더 했음. ( 홈/ 어웨이 기준으로 팀을 묶어버림 )

이 데이터를 통해 기술 통계를 나타내는 것은 부정확할 수 있다고 생각.

- 따라서 가공 하기전 팀 별 데이터가 남아 있는 'K-League-data.csv'를 기준으로 기술 통계를 나타내고자 함. 



----------------

In [16]:
import pandas as pd
df = pd.read_csv('K-League-data.csv')
df = df.drop(columns='Unnamed: 0')
df.info()

# 행과 열 : ( 133, 52 ) , 결측치 없음 , dtypes는 3가지 : 소수 19개, 정수 29개 , object 4개 (대회, H/A, 기준, 경기결과)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1332 entries, 0 to 1331
Data columns (total 52 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Rnd.      1332 non-null   int64  
 1   대회        1332 non-null   object 
 2   H/A       1332 non-null   object 
 3   득점        1332 non-null   int64  
 4   도움        1332 non-null   int64  
 5   슈팅        1332 non-null   int64  
 6   유효 슈팅     1332 non-null   int64  
 7   블락된슈팅     1332 non-null   int64  
 8   벗어난슈팅     1332 non-null   int64  
 9   PA내 슈팅    1332 non-null   int64  
 10  PA외 슈팅    1332 non-null   int64  
 11  프리킥 슈팅    1332 non-null   int64  
 12  프리킥 유효슈팅  1332 non-null   int64  
 13  프리킥 크로스   1332 non-null   float64
 14  오프사이드     1332 non-null   int64  
 15  코너킥       1332 non-null   int64  
 16  스로인       1332 non-null   int64  
 17  드리블       1332 non-null   float64
 18  패스        1332 non-null   float64
 19  키패스       1332 non-null   int64  
 20  전방패스      1332 non-null   floa

In [108]:
# 기술통계량 1

df[['득점', '도움', '슈팅', '유효 슈팅', '블락된슈팅', '벗어난슈팅','PA내 슈팅', 'PA외 슈팅', '프리킥 슈팅',
    '프리킥 유효슈팅', '프리킥 크로스', '오프사이드', '코너킥','스로인', '드리블', '패스']].describe()

득점           도움           슈팅        유효 슈팅        블락된슈팅  \
count  1332.000000  1332.000000  1332.000000  1332.000000  1332.000000   
mean      1.263514     0.795796    11.460961     4.007508     2.478979   
std       1.098632     0.897637     4.053268     2.174562     1.823631   
min       0.000000     0.000000     1.000000     0.000000     0.000000   
25%       0.000000     0.000000     9.000000     2.000000     1.000000   
50%       1.000000     1.000000    11.000000     4.000000     2.000000   
75%       2.000000     1.000000    14.000000     5.000000     3.000000   
max       7.000000     5.000000    27.000000    13.000000    11.000000   

             벗어난슈팅       PA내 슈팅       PA외 슈팅       프리킥 슈팅     프리킥 유효슈팅  \
count  1332.000000  1332.000000  1332.000000  1332.000000  1332.000000   
mean      4.974474     6.593844     4.867117     0.431682     0.139640   
std       2.367310     2.982452     2.455231     0.652558     0.377849   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       3.000000     4.000000     3.000000     0.000000     0.000000   
50%       5.000000     6.000000     5.000000     0.000000     0.000000   
75%       6.000000     8.000000     6.000000     1.000000     0.000000   
max      15.000000    20.000000    16.000000     4.000000     2.000000   

           프리킥 크로스        오프사이드          코너킥          스로인          드리블  \
count  1332.000000  1332.000000  1332.000000  1332.000000  1332.000000   
mean     19.920195     1.408408     4.290541    19.987237    36.894444   
std      32.584450     1.324742     2.396241     5.900807    24.104584   
min       0.000000     0.000000     0.000000     7.000000     0.000000   
25%       0.000000     0.000000     3.000000    16.000000    20.000000   
50%       0.000000     1.000000     4.000000    20.000000    36.400000   
75%      42.500000     2.000000     6.000000    24.000000    50.000000   
max     100.000000    10.000000    17.000000    43.000000   100.000000   

                패스  
count  1332.000000  
mean     81.009685  
std       6.474343  
min      53.500000  
25%      76.900000  
50%      81.900000  
75%      86.000000  
max      94.300000

In [109]:
# 기술통계량 2

df[['키패스', '전방패스', '후방패스', '횡패스', '공격진영 패스', '수비진영 패스',
       '중앙진영 패스', '롱패스', '중거리패스', '단거리패스', '크로스', '탈압박', '경합 지상', '경합 공중',
       '태클', '클리어링', '인터셉트', '차단', '획득']].describe()

키패스         전방패스         후방패스          횡패스      공격진영 패스  \
count  1332.000000  1332.000000  1332.000000  1332.000000  1332.000000   
mean      6.752252    66.447973    95.989865    89.623724    75.637312   
std       2.961339     8.372458     2.569367     5.567650     8.765133   
min       0.000000    39.600000    81.800000    64.800000    41.300000   
25%       5.000000    60.700000    94.600000    86.500000    69.400000   
50%       7.000000    67.000000    96.400000    90.900000    76.650000   
75%       9.000000    72.625000    97.800000    93.900000    82.500000   
max      18.000000    87.000000   100.000000   100.000000    97.800000   

           수비진영 패스      중앙진영 패스          롱패스        중거리패스        단거리패스  \
count  1332.000000  1332.000000  1332.000000  1332.000000  1332.000000   
mean     77.731532    84.567718    56.392342    85.203679    85.058934   
std       9.650162     5.999876    12.315111     6.340819     5.190515   
min      39.600000    56.600000    16.000000    59.600000    57.800000   
25%      71.675000    81.000000    47.900000    81.400000    82.100000   
50%      79.200000    85.600000    55.600000    86.300000    85.800000   
75%      85.125000    89.000000    63.800000    90.000000    88.900000   
max      96.500000    97.100000    90.800000    96.800000    95.400000   

               크로스          탈압박        경합 지상       경합 공중           태클  \
count  1332.000000  1332.000000  1332.000000  1332.00000  1332.000000   
mean     22.128378     2.578078    49.995420    50.00000    39.943393   
std      10.312541     2.281286    10.477355     8.49709    18.117761   
min       0.000000     0.000000     0.000000    23.80000     0.000000   
25%      15.400000     1.000000    44.400000    44.20000    27.600000   
50%      21.400000     2.000000    50.000000    50.00000    37.500000   
75%      28.600000     4.000000    55.600000    55.80000    50.000000   
max      62.500000    14.000000   100.000000    76.20000   100.000000   

              클리어링         인터셉트          차단           획득  
count  1332.000000  1332.000000  1332.00000  1332.000000  
mean     27.536036    26.565315    25.37988    63.234234  
std      13.753690     9.359556    12.52459    18.389494  
min       2.000000     3.000000     5.00000    25.000000  
25%      17.000000    20.000000    16.00000    49.000000  
50%      26.000000    27.000000    21.00000    60.000000  
75%      37.000000    33.000000    34.00000    76.000000  
max      74.000000    64.000000    72.00000   125.000000

In [110]:
# 기술통계량 3

df[['블락', '볼미스', '파울', '피파울', '경고', '퇴장',
       '시즌', '경기결과', '소속 선수 수', '평균 나이', '용병 수', '평균 가치', '총 가치']].describe()

블락          볼미스           파울          피파울           경고  \
count  1332.000000  1332.000000  1332.000000  1332.000000  1332.000000   
mean     14.014264     6.609610    11.750000    11.277027     1.748498   
std       9.583307     2.786546     3.891055     3.774323     1.278836   
min       0.000000     0.000000     2.000000     2.000000     0.000000   
25%       4.000000     5.000000     9.000000     9.000000     1.000000   
50%      15.000000     6.000000    11.000000    11.000000     2.000000   
75%      21.000000     8.000000    14.000000    14.000000     3.000000   
max      44.000000    17.000000    27.000000    25.000000     8.000000   

                퇴장           시즌      소속 선수 수        평균 나이         용병 수  \
count  1332.000000  1332.000000  1332.000000  1332.000000  1332.000000   
mean      0.032282  2021.972973    39.423423    24.899925     4.484234   
std       0.181014     0.810815     5.294142     1.376586     1.282919   
min       0.000000  2021.000000    30.000000    22.600000     0.000000   
25%       0.000000  2021.000000    36.000000    23.900000     4.000000   
50%       0.000000  2022.000000    40.000000    24.800000     5.000000   
75%       0.000000  2023.000000    42.000000    25.900000     5.000000   
max       2.000000  2023.000000    52.000000    28.800000     7.000000   

             평균 가치         총 가치  
count  1332.000000  1332.000000  
mean    351.012012    13.684707  
std      98.703754     3.590314  
min     228.000000     9.180000  
25%     287.000000    11.030000  
50%     322.000000    12.230000  
75%     378.000000    17.080000  
max     632.000000    22.180000

In [122]:
# 부족한 기술통계량 보기 // 최빈값
# 범주형 데이터 삭제.
# 삭제한 것 : Rnd. 대회. 기준. 시즌.


df_columns = list(df.columns)
df_columns.remove('기준')
df_columns.remove('시즌')
df_columns.remove('경기결과')

for column in df_columns[3:]:
    # print(column)
    print(column, df[column].mode()[0])

득점 1
도움 0
슈팅 9
유효 슈팅 4
블락된슈팅 2
벗어난슈팅 5
PA내 슈팅 7
PA외 슈팅 4
프리킥 슈팅 0
프리킥 유효슈팅 0
프리킥 크로스 0.0
오프사이드 1
코너킥 3
스로인 18
드리블 0.0
패스 82.3
키패스 7
전방패스 66.8
후방패스 100.0
횡패스 94.9
공격진영 패스 66.7
수비진영 패스 75.9
중앙진영 패스 88.8
롱패스 50.0
중거리패스 88.2
단거리패스 87.1
크로스 25.0
탈압박 1
경합 지상 50.0
경합 공중 50.0
태클 50.0
클리어링 18
인터셉트 25
차단 18
획득 49
블락 2
볼미스 6
파울 10
피파울 9
경고 1
퇴장 0
소속 선수 수 40
평균 나이 26.3
용병 수 4
평균 가치 228.0
총 가치 15.05


In [134]:
# 부족한 기술통게량 보기 // 최빈값
# 범주형 데이터 ( dtype이 object이므로 mode연산 불가 ) 
# value_count()를 통해 최빈값 찾기

df['Rnd.'].value_counts()

1     36
19    36
21    36
22    36
23    36
24    36
25    36
26    36
27    36
28    36
29    36
30    36
31    36
32    36
33    36
34    36
35    36
2     36
20    36
18    36
17    36
3     36
4     36
5     36
6     36
7     36
8     36
9     36
10    36
11    36
12    36
13    36
14    36
15    36
16    36
36    24
37    24
38    24
Name: Rnd., dtype: int64

In [131]:
df['대회'].value_counts()

전북vs포항      14
대구vs수원FC    12
수원FCvs서울    12
제주vs전북      12
강원vs인천      12
            ..
대구vs대전       2
대전vs포항       2
대전vs인천       2
김천vs수원FC     2
제주vs김천       2
Name: 대회, Length: 176, dtype: int64

In [132]:
df['경기결과'].value_counts()

L    469
W    469
D    394
Name: 경기결과, dtype: int64

In [135]:
df['시즌'].value_counts()

2021    456
2022    456
2023    420
Name: 시즌, dtype: int64

In [136]:
df['기준'].value_counts()

daegu       111
jeju        111
seoul       111
gangwon     111
ulsan       111
suwonFC     111
incheon     111
jeonbuk     111
suwon       111
pohang      111
seongnam     76
gwangju      73
gimcheon     38
daejeon      35
Name: 기준, dtype: int64

--------

## 변수 간 상관관계 분석

In [56]:
# 변수간의 상관관계 분석 
# 나눠서 시각화 1

df[['득점', '도움', '슈팅', '유효 슈팅', '블락된슈팅', '벗어난슈팅','PA내 슈팅', 'PA외 슈팅', '프리킥 슈팅',
    '프리킥 유효슈팅', '프리킥 크로스', '오프사이드', '코너킥','스로인', '드리블', '패스']].corr()

득점        도움        슈팅     유효 슈팅     블락된슈팅     벗어난슈팅  \
득점        1.000000  0.791314  0.239278  0.551404 -0.067921 -0.044499   
도움        0.791314  1.000000  0.163832  0.424562 -0.082485 -0.045943   
슈팅        0.239278  0.163832  1.000000  0.614957  0.571126  0.707336   
유효 슈팅     0.551404  0.424562  0.614957  1.000000  0.051004  0.095049   
블락된슈팅    -0.067921 -0.082485  0.571126  0.051004  1.000000  0.160681   
벗어난슈팅    -0.044499 -0.045943  0.707336  0.095049  0.160681  1.000000   
PA내 슈팅    0.312200  0.221010  0.798096  0.580274  0.392743  0.530913   
PA외 슈팅    0.015777  0.001996  0.681397  0.310335  0.465777  0.522802   
프리킥 슈팅   -0.003689 -0.034096  0.204218  0.091958  0.167045  0.136507   
프리킥 유효슈팅  0.095899  0.039833  0.093826  0.194403 -0.004460 -0.014491   
프리킥 크로스   0.033110  0.045096  0.053264  0.048672  0.020859  0.030420   
오프사이드    -0.009473 -0.011318 -0.051038  0.012236 -0.074503 -0.041234   
코너킥      -0.026821 -0.044351  0.383262  0.208217  0.216227  0.298383   
스로인      -0.120242 -0.094109  0.032601 -0.060652  0.068223  0.058978   
드리블       0.074863  0.042577  0.128674  0.106596  0.080585  0.060319   
패스        0.007700  0.027360  0.255120  0.130674  0.085335  0.251041   

            PA내 슈팅    PA외 슈팅    프리킥 슈팅  프리킥 유효슈팅   프리킥 크로스     오프사이드  \
득점        0.312200  0.015777 -0.003689  0.095899  0.033110 -0.009473   
도움        0.221010  0.001996 -0.034096  0.039833  0.045096 -0.011318   
슈팅        0.798096  0.681397  0.204218  0.093826  0.053264 -0.051038   
유효 슈팅     0.580274  0.310335  0.091958  0.194403  0.048672  0.012236   
블락된슈팅     0.392743  0.465777  0.167045 -0.004460  0.020859 -0.074503   
벗어난슈팅     0.530913  0.522802  0.136507 -0.014491  0.030420 -0.041234   
PA내 슈팅    1.000000  0.102819  0.029933 -0.006970  0.105992 -0.000390   
PA외 슈팅    0.102819  1.000000  0.300777  0.163362 -0.040821 -0.083784   
프리킥 슈팅    0.029933  0.300777  1.000000  0.541488 -0.000774 -0.047657   
프리킥 유효슈팅 -0.006970  0.163362  0.541488  1.000000 -0.011354 -0.032967   
프리킥 크로스   0.105992 -0.040821 -0.000774 -0.011354  1.000000  0.000094   
오프사이드    -0.000390 -0.083784 -0.047657 -0.032967  0.000094  1.000000   
코너킥       0.356298  0.199909  0.078769  0.015733 -0.036013  0.020342   
스로인       0.045000 -0.000843  0.035382 -0.002233 -0.035748  0.038247   
드리블       0.139178  0.043360 -0.055693 -0.033323 -0.030255 -0.008726   
패스        0.194272  0.185182 -0.000083 -0.023065 -0.005798 -0.049569   

               코너킥       스로인       드리블        패스  
득점       -0.026821 -0.120242  0.074863  0.007700  
도움       -0.044351 -0.094109  0.042577  0.027360  
슈팅        0.383262  0.032601  0.128674  0.255120  
유효 슈팅     0.208217 -0.060652  0.106596  0.130674  
블락된슈팅     0.216227  0.068223  0.080585  0.085335  
벗어난슈팅     0.298383  0.058978  0.060319  0.251041  
PA내 슈팅    0.356298  0.045000  0.139178  0.194272  
PA외 슈팅    0.199909 -0.000843  0.043360  0.185182  
프리킥 슈팅    0.078769  0.035382 -0.055693 -0.000083  
프리킥 유효슈팅  0.015733 -0.002233 -0.033323 -0.023065  
프리킥 크로스  -0.036013 -0.035748 -0.030255 -0.005798  
오프사이드     0.020342  0.038247 -0.008726 -0.049569  
코너킥       1.000000  0.165459  0.065981  0.215090  
스로인       0.165459  1.000000 -0.061389 -0.031441  
드리블       0.065981 -0.061389  1.000000  0.024534  
패스        0.215090 -0.031441  0.024534  1.000000

In [57]:
# 변수간의 상관관계 분석 
# 나눠서 시각화 2

df[['키패스', '전방패스', '후방패스', '횡패스', '공격진영 패스', '수비진영 패스',
       '중앙진영 패스', '롱패스', '중거리패스', '단거리패스', '크로스', '탈압박', '경합 지상', '경합 공중',
       '태클', '클리어링', '인터셉트', '차단', '획득']].corr()

키패스      전방패스      후방패스       횡패스   공격진영 패스   수비진영 패스   중앙진영 패스  \
키패스      1.000000  0.280364  0.120334  0.098743  0.195195  0.205721  0.220792   
전방패스     0.280364  1.000000  0.303881  0.574986  0.445892  0.827072  0.832984   
후방패스     0.120334  0.303881  1.000000  0.310881  0.294148  0.360370  0.408247   
횡패스      0.098743  0.574986  0.310881  1.000000  0.693175  0.624607  0.648669   
공격진영 패스  0.195195  0.445892  0.294148  0.693175  1.000000  0.365085  0.395567   
수비진영 패스  0.205721  0.827072  0.360370  0.624607  0.365085  1.000000  0.691785   
중앙진영 패스  0.220792  0.832984  0.408247  0.648669  0.395567  0.691785  1.000000   
롱패스      0.279448  0.585652  0.262526  0.432995  0.387632  0.680024  0.612370   
중거리패스    0.172259  0.780361  0.362586  0.769226  0.560769  0.764918  0.812665   
단거리패스    0.182541  0.786733  0.484110  0.762213  0.611046  0.706302  0.824635   
크로스      0.402053  0.024966 -0.032452  0.051403  0.080993  0.010580 -0.004648   
탈압박      0.031670  0.138410  0.088341  0.173548  0.181142  0.155442  0.115795   
경합 지상    0.034930  0.089454  0.051265  0.030857  0.045125  0.059280  0.067144   
경합 공중    0.159971  0.150181  0.034183  0.033500 -0.021489  0.197886  0.061513   
태클       0.038985 -0.056226 -0.046341 -0.276195 -0.307213 -0.110095 -0.030542   
클리어링    -0.132817 -0.198979 -0.082221  0.194995  0.324434 -0.201962 -0.238952   
인터셉트     0.010451  0.049605  0.047773  0.356889  0.446802  0.091576  0.033825   
차단      -0.052478 -0.315203 -0.193808 -0.594894 -0.651417 -0.341339 -0.297754   
획득       0.132297 -0.174922 -0.103481 -0.497414 -0.572103 -0.200502 -0.180602   

              롱패스     중거리패스     단거리패스       크로스       탈압박     경합 지상     경합 공중  \
키패스      0.279448  0.172259  0.182541  0.402053  0.031670  0.034930  0.159971   
전방패스     0.585652  0.780361  0.786733  0.024966  0.138410  0.089454  0.150181   
후방패스     0.262526  0.362586  0.484110 -0.032452  0.088341  0.051265  0.034183   
횡패스      0.432995  0.769226  0.762213  0.051403  0.173548  0.030857  0.033500   
공격진영 패스  0.387632  0.560769  0.611046  0.080993  0.181142  0.045125 -0.021489   
수비진영 패스  0.680024  0.764918  0.706302  0.010580  0.155442  0.059280  0.197886   
중앙진영 패스  0.612370  0.812665  0.824635 -0.004648  0.115795  0.067144  0.061513   
롱패스      1.000000  0.553581  0.435648  0.061634  0.128664  0.039840  0.192747   
중거리패스    0.553581  1.000000  0.712334  0.031898  0.143382  0.059859  0.098576   
단거리패스    0.435648  0.712334  1.000000 -0.000764  0.160416  0.086671  0.004377   
크로스      0.061634  0.031898 -0.000764  1.000000  0.005424  0.003229  0.160649   
탈압박      0.128664  0.143382  0.160416  0.005424  1.000000  0.028674 -0.025511   
경합 지상    0.039840  0.059859  0.086671  0.003229  0.028674  1.000000 -0.019790   
경합 공중    0.192747  0.098576  0.004377  0.160649 -0.025511 -0.019790  1.000000   
태클      -0.109083 -0.154181 -0.130236 -0.035717 -0.074443  0.084854  0.006917   
클리어링    -0.207917 -0.034692 -0.062386  0.010860  0.040905 -0.001599 -0.007810   
인터셉트     0.131708  0.178609  0.145298 -0.006977  0.199115  0.015949  0.022715   
차단      -0.334409 -0.469823 -0.423008 -0.029547 -0.213006  0.038191  0.008559   
획득      -0.243495 -0.366329 -0.322539 -0.063544 -0.125889 -0.008778  0.023032   

               태클      클리어링      인터셉트        차단        획득  
키패스      0.038985 -0.132817  0.010451 -0.052478  0.132297  
전방패스    -0.056226 -0.198979  0.049605 -0.315203 -0.174922  
후방패스    -0.046341 -0.082221  0.047773 -0.193808 -0.103481  
횡패스     -0.276195  0.194995  0.356889 -0.594894 -0.497414  
공격진영 패스 -0.307213  0.324434  0.446802 -0.651417 -0.572103  
수비진영 패스 -0.110095 -0.201962  0.091576 -0.341339 -0.200502  
중앙진영 패스 -0.030542 -0.238952  0.033825 -0.297754 -0.180602  
롱패스     -0.109083 -0.207917  0.131708 -0.334409 -0.243495  
중거리패스   -0.154181 -0.034692  0.178609 -0.469823 -0.366329  
단거리패스   -0.130236 -0.062386  0.145298 -0.423008 -0.322539  
크로스     -0.035717  0.010860 -0.006977 -0.029547 -0.063544  
탈압박     -0.074443  0.040905  0.199115 -0.213006 -0.1

In [61]:
# 변수간의 상관관계 분석 
# 나눠서 시각화 2

df[['블락', '볼미스', '파울', '피파울', '경고', '퇴장',
       '시즌', '경기결과', '소속 선수 수', '평균 나이', '용병 수', '평균 가치', '총 가치']].corr()

C:\Users\HONG\AppData\Local\Temp\ipykernel_4964\1758729795.py:4: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df[['블락', '볼미스', '파울', '피파울', '경고', '퇴장',


블락       볼미스        파울       피파울        경고        퇴장        시즌  \
블락       1.000000  0.077691 -0.176323 -0.231815  0.037811 -0.010660  0.693129   
볼미스      0.077691  1.000000 -0.012542 -0.007711  0.014172 -0.019681  0.133660   
파울      -0.176323 -0.012542  1.000000  0.116705  0.332511  0.000800 -0.252667   
피파울     -0.231815 -0.007711  0.116705  1.000000  0.068147  0.064978 -0.244039   
경고       0.037811  0.014172  0.332511  0.068147  1.000000 -0.023321  0.033291   
퇴장      -0.010660 -0.019681  0.000800  0.064978 -0.023321  1.000000 -0.029884   
시즌       0.693129  0.133660 -0.252667 -0.244039  0.033291 -0.029884  1.000000   
소속 선수 수  0.054509  0.014677  0.113245 -0.069118  0.046702  0.031981  0.001618   
평균 나이    0.153131  0.030665 -0.198690 -0.039762 -0.013497 -0.017478  0.460350   
용병 수     0.056880  0.026649  0.020657 -0.071946  0.041315  0.000576  0.231439   
평균 가치   -0.178828 -0.092115 -0.078144  0.001165 -0.060021 -0.035723 -0.110791   
총 가치    -0.133704 -0.086423 -0.029319 -0.039360 -0.044769 -0.022973 -0.115045   

          소속 선수 수     평균 나이      용병 수     평균 가치      총 가치  
블락       0.054509  0.153131  0.056880 -0.178828 -0.133704  
볼미스      0.014677  0.030665  0.026649 -0.092115 -0.086423  
파울       0.113245 -0.198690  0.020657 -0.078144 -0.029319  
피파울     -0.069118 -0.039762 -0.071946  0.001165 -0.039360  
경고       0.046702 -0.013497  0.041315 -0.060021 -0.044769  
퇴장       0.031981 -0.017478  0.000576 -0.035723 -0.022973  
시즌       0.001618  0.460350  0.231439 -0.110791 -0.115045  
소속 선수 수  1.000000 -0.643184  0.279410 -0.298139  0.241358  
평균 나이   -0.643184  1.000000  0.026269  0.323416 -0.037918  
용병 수     0.279410  0.026269  1.000000 -0.081983  0.032105  
평균 가치   -0.298139  0.323416 -0.081983  1.000000  0.844729  
총 가치     0.241358 -0.037918  0.032105  0.844729  1.000000

In [21]:
# 팀별 승리/무승부/패배 개수

grouped_data = df.groupby(['기준', '경기결과'])

result_counts = df.groupby(['기준', '경기결과']).size().reset_index(name='개수')
print(result_counts)



          기준 경기결과  개수
0      daegu    D  39
1      daegu    L  34
2      daegu    W  38
3    daejeon    D  14
4    daejeon    L  10
5    daejeon    W  11
6    gangwon    D  35
7    gangwon    L  48
8    gangwon    W  28
9   gimcheon    D  14
10  gimcheon    L  16
11  gimcheon    W   8
12   gwangju    D  17
13   gwangju    L  32
14   gwangju    W  24
15   incheon    D  38
16   incheon    L  34
17   incheon    W  39
18      jeju    D  35
19      jeju    L  40
20      jeju    W  36
21   jeonbuk    D  28
22   jeonbuk    L  25
23   jeonbuk    W  58
24    pohang    D  36
25    pohang    L  32
26    pohang    W  43
27  seongnam    D  19
28  seongnam    L  38
29  seongnam    W  19
30     seoul    D  34
31     seoul    L  40
32     seoul    W  37
33     suwon    D  30
34     suwon    L  51
35     suwon    W  30
36   suwonFC    D  27
37   suwonFC    L  50
38   suwonFC    W  34
39     ulsan    D  28
40     ulsan    L  19
41     ulsan    W  64


- 변수 선택법으로 선택된 변수들의 기술통계 분석

 승리 / 무승부 / 패배 로 나누어봤음

In [29]:
# 선택된 변수들
origin_selected_features = ['홈 득점', '원정 득점', '홈 경고', '원정 팀_daejeon']

# 원정 팀_daejeon은 기술통계에서 제외
selected_features = ['홈 득점', '원정 득점', '홈 경고']

# 재가공한 데이터프레임 new_df에서 기술통계
new_df[selected_features]

홈 득점  원정 득점  홈 경고
0       1      1     1
1       0      0     1
2       3      1     2
3       2      2     2
4       0      0     3
..    ...    ...   ...
661     0      0     0
662     2      0     2
663     2      2     3
664     0      1     1
665     1      0     2

[666 rows x 3 columns]

In [44]:
import pandas as pd

# 필요한 열만 추출
selected_columns = ['홈 득점', '원정 득점', '홈 경고', '홈 경기결과']
selected_data = new_df[selected_columns]

# 승/무/패 각각의 데이터 추출
winning_data = selected_data[selected_data['홈 경기결과'] == 'W']
draw_data = selected_data[selected_data['홈 경기결과'] == 'D']
losing_data = selected_data[selected_data['홈 경기결과'] == 'L']

# 기술통계 계산
# winning_describe = winning_data[['홈 득점', '원정 득점', '홈 경고']].describe()
# draw_describe = draw_data[['홈 득점', '원정 득점', '홈 경고']].describe()
# losing_describe = losing_data[['홈 득점', '원정 득점', '홈 경고']].describe()
winning_describe = winning_data.describe()
draw_describe = draw_data.describe()
losing_describe = losing_data.describe()

# 출력
print("승리에 대한 기술통계:")
print(winning_describe)

print("\n무승부에 대한 기술통계:")
print(draw_describe)

print("\n패배에 대한 기술통계:")
print(losing_describe)


승리에 대한 기술통계:
             홈 득점       원정 득점        홈 경고
count  260.000000  260.000000  260.000000
mean     2.134615    0.580769    1.753846
std      1.036568    0.743839    1.300796
min      0.000000    0.000000    0.000000
25%      1.000000    0.000000    1.000000
50%      2.000000    0.000000    2.000000
75%      3.000000    1.000000    3.000000
max      7.000000    3.000000    7.000000

무승부에 대한 기술통계:
             홈 득점       원정 득점        홈 경고
count  197.000000  197.000000  197.000000
mean     0.994924    0.989848    1.472081
std      0.792165    0.788889    1.122807
min      0.000000    0.000000    0.000000
25%      0.000000    0.000000    1.000000
50%      1.000000    1.000000    1.000000
75%      2.000000    1.000000    2.000000
max      4.000000    4.000000    5.000000

패배에 대한 기술통계:
             홈 득점       원정 득점        홈 경고
count  209.000000  209.000000  209.000000
mean     0.660287    2.143541    1.746411
std      0.799280    1.032393    1.223881
min      0.000000    1.000000    0

# 최적변수 관련 설명

변수선택법에 3가지 종류가 있음

- 전진 선택법 : 기존 모형에 가장 설명력이 좋은 변수를 하나씩 추가해가는 방법

- 후진 소거법 : 모든 변수가 포함된 모형에서 설명력이 안좋은 변수를 하나씩 제거해 나가는 방법

- 단계적 선택법 : 전진 선택법에서 후진 소거법을 추가한 방법

일반적으로 단계적 선택법이 좋음. 키 페이퍼도 단계적 선택법을 사용하였음.

단계 선택법이란 ? 최소 p 값과 유의 수준, 최종적으로 최대 p 값까지 고려하여 변수를 추가/제거 하는 방법 


# 겪은 시행착오


## 1.

처음 수집한 데이터를 df로 만들었을 때, (1332, 52) 의 쉐잎이었음.

한 라운드의, 1개의 매치(컬럼명 :대회)에 2개의 행을 가지고 있는 형식

ex) 1라운드/  포항vs제주 /  - 포항 데이터 -

ex) 1라운드/  제주vs포항 /  - 제주 데이터 - 

따라서 대회 변수, H/A 변수를 더미변수로 만들었음

이 과정에서 더미변수가 너무 많아져버림. 구단과 구단 사이에 수많은 경우의 수가 0 or 1로 구성되었다.

이 데이터프레임 형식일때 모델의 정확도는 전체적으로 60% 정도

- 이 df로 변수 선택법 하니 225개중 50여개정도. 성능이 좋아지지도 않았음

------------

더미 변수가 너무 많아 학습력이 떨어진다고 생각했음 ( 0이 너무 많았다고 생각, 열도 너무 늘어났음 )

더미 변수를 줄이기 위해, 수집한 데이터의 형식을 바꿔버림

기존 한개의 라운드, 1개의 매치에 2개의 행을 가지고 있는 형식 >>> 한개의 라운드, 1개의 매치에 1개의 행을 가지고 있는 형식

ex) 1라운드 / 포항 vs 제주 / - 포항(홈) 데이터 - / - 제주(원정) 데이터 - 

이에 따라 데이터 쉐잎도 바뀜 ( 662,100 )  

더미 변수 개수가 이전엔 200여개였는데 100여개로 줄었음

이 데이터 프레임 형식일때의 모델의 정확도는 전체적으로 80% 이상이다.

- 이 df로 변수 선택법 하니 4개. 성능 오름

## 2.

PCA랑 변수 선택법이랑 동일한 방식인줄 알았음. 

하지만 아니었다.